src/meta/src/stream/fragmenter/mod.rs

```diff
@@ -29,51 +31,41 @@ use risingwave_pb::stream_plan::{
     DispatchStrategy, Dispatcher, DispatcherType, ExchangeNode, StreamNode,
 };
 
-use super::graph::StreamFragmentEdge;
 use super::{CreateMaterializedViewContext, FragmentManagerRef};
 use crate::manager::{IdCategory, IdGeneratorManagerRef};
 use crate::model::{FragmentId, LocalActorId, LocalFragmentId};
 use crate::storage::MetaStore;
-use crate::stream::graph::{
-    StreamActorBuilder, StreamFragment, StreamFragmentGraph, StreamGraphBuilder,
-};
 
 /// [`StreamFragmenter`] generates the proto for interconnected actors for a streaming pipeline.
 pub struct StreamFragmenter<S> {
+    // ==== Local states ====
     /// fragment graph field, transformed from input streaming plan.
-    pub(super) fragment_graph: StreamFragmentGraph,
-
-    /// fragment manager, used to retrieve upstream table fragment infos (dist_keys, actor_ids).
-    pub(super) fragment_manager: FragmentManagerRef<S>,
-
+    fragment_graph: StreamFragmentGraph,
     /// stream graph builder, to build streaming DAG.
-    stream_graph: StreamGraphBuilder,
-
-    /// id generator, used to generate actor id.
-    id_gen_manager: IdGeneratorManagerRef<S>,
+    stream_graph_builder: StreamGraphBuilder,
+    /// when converting fragment graph to actor graph, we need to know which actors belong to a
+    /// fragment.
+    fragment_actors: HashMap<LocalFragmentId, Vec<LocalActorId>>,
+    /// dependent table ids
+    dependent_table_ids: HashSet<TableId>,
+    /// upstream dispatch keys
+    distribution_keys: Vec<i32>,
```

Comment:
There should be multiple sets of distribution keys, one set per upstream?

Reply:
cc @zbzbw why there's only one set of distribution keys? 🤣

Reply:
> cc @zbzbw why there's only one set of distribution keys? 🤣

My fault😢. The comment was misleading, this field contains current fragment's distribution key instead of upstream.

Reply:
Let's fix this later 🥲